File: 05-DecisionesSimples/04-Donar/02-Go/Donar.go
/*
   Programa para verificar la elegibilidad de los donantes
   con base en la edad, el peso y la madurez fisiológica
   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Mayo 2025
   Licencia: GNU GPL v3
*/

package main

import (
	"donar/modulo"
)

const (
	EDAD_MINIMA              = 18
	EDAD_MAXIMA              = 65
	EDAD_MINIMA_AUTORIZACION = 16
	EDAD_MAXIMA_AUTORIZACION = 18
	PESO_MINIMO              = 50.0
)

func main() {
	edadDonante := modulo.IngresarEntero("Edad del donante: ")
	autorizacionEdad := modulo.IngresarLogico(
		"Tiene autorización por edad (s/n): ")
	pesoDonante := modulo.IngresarReal("Peso del donante: ")
	suficienteMadurez := modulo.IngresarLogico(
		"Suficiente madurez fisiológica(s/n): ")

	reporteLegibilidad := generarReporteEligibilidad(
		edadDonante, autorizacionEdad, pesoDonante, suficienteMadurez)

	modulo.MostrarMensaje(reporteLegibilidad)
}

func generarReporteEligibilidad(
	edadDonante int, autorizacionEdad bool, pesoDonante float64, suficienteMadurez bool) string {
	mensaje := "\nEl donante no cumple las condiciones para donar\n"

	if (edadDonante >= EDAD_MINIMA && edadDonante <= EDAD_MAXIMA ||
		edadDonante >= EDAD_MINIMA_AUTORIZACION &&
			edadDonante <= EDAD_MAXIMA_AUTORIZACION && autorizacionEdad) &&
		pesoDonante > PESO_MINIMO && suficienteMadurez {
		mensaje = "\nEl donante es elegible para donar\n"
	}

	return mensaje
}

File: 05-DecisionesSimples/04-Donar/05-Haskell/Donar.hs
{-
    Programa para verificar la elegibilidad de los donantes 
    con base en la edad, el peso y la madurez fisiológica
    Autor: Julián Esteban Gutiérrez Posada
    Fecha: Mayo 2025
    Licencia: GNU GPL v3
-}

import Modulo.Util (mostrarMensaje, ingresarEntero, ingresarLogico, ingresarReal)

edad_minima, edad_maxima, edad_minima_autorizacion, edad_maxima_autorizacion :: Int
edad_minima = 18
edad_maxima = 65
edad_minima_autorizacion = 16
edad_maxima_autorizacion = 18

peso_minimo :: Double
peso_minimo = 50.0

main :: IO()
main = do
  edadDonante <- ingresarEntero "Edad del donante: "
  autorizacionEdad <- ingresarLogico "Tiene autorización por edad (s/n): "
  pesoDonante <- ingresarReal "Peso del donante: "
  suficienteMadurez <-ingresarLogico "Suficiente madurez fisiológica(s/n):"

  let reporteLegibilidad = generarReporteEligibilidad edadDonante autorizacionEdad pesoDonante suficienteMadurez  

  mostrarMensaje reporteLegibilidad

generarReporteEligibilidad :: Int -> Bool -> Double -> Bool -> String
generarReporteEligibilidad edadDonante autorizacionEdad pesoDonante suficienteMadurez =
    if (edadDonante >= edad_minima && edadDonante <= edad_maxima ||
      edadDonante >= edad_minima_autorizacion &&
      edadDonante <= edad_maxima_autorizacion && autorizacionEdad) &&
      pesoDonante > peso_minimo && suficienteMadurez
    then "\nEl donante es elegible para donar\n"
    else "\nEl donante no cumple las condiciones para donar\n"